pkg/types/validation/installconfig.go

```diff
@@ -1053,8 +1054,29 @@ func validateFeatureSet(c *types.InstallConfig) field.ErrorList {
 		allErrs = append(allErrs, validateCustomFeatureGates(c)...)
 	}
 
-	if c.FeatureSet != configv1.TechPreviewNoUpgrade {
-		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
+	if c.FeatureSet == configv1.CustomNoUpgrade && len(c.FeatureGates) > 0 {
+		features, err := featuregates.GenerateCustomFeatures(c.FeatureGates)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("featureGates"), c.FeatureGates, err.Error()))
+			return allErrs
+		}
+		featureGates := featuregates.NewFeatureGate(features.Enabled, features.Disabled)
+
+		if c.GCP != nil {
+			if !featureGates.Enabled(configv1.FeatureGateGCPLabelsTags) {
+				errMsg := "featureGate GCPLabelsTags must be enabled with CustomNoUpgrade featureSet to use this field"
+				if len(c.GCP.UserTags) > 0 {
+					allErrs = append(allErrs, field.Forbidden(field.NewPath("platform", "gcp", "userTags"), errMsg))
+				}
+				if len(c.GCP.UserLabels) > 0 {
+					allErrs = append(allErrs, field.Forbidden(field.NewPath("platform", "gcp", "userLabels"), errMsg))
+				}
+			}
+		}
+	}
+
+	if c.FeatureSet != configv1.TechPreviewNoUpgrade && c.FeatureSet != configv1.CustomNoUpgrade {
```

Comment:
We shouldn't need separate checks for TechPreviewNoUpgrade or CustomNoUpgrade. Either of these feature sets should populate the feature gates, and then we only need a single check against the feature gates. Currently this doesn't work, though, because featureGates is not being populated when TechPreview is set.

Reply:
Simplified the code by moving GCP tag feature checks to gcp package.

Reply:
> Currently this doesn't work, though, because featureGates is not being populated when TechPreview is set.

Was this addressed? Is my concern here clear/accurate?

Reply:
@patrickdillon I think I misunderstood the feedback and I assumed if TechPreviewNoUpgrade is set, then all features marked for TechPreview are by default enabled and updated the code accordingly.

Please correct me, suggestion is to populate featureGates for both TechPreviewNoUpgrade and CustomNoUpgrade, and to make use of featureGates instead of featureSets for checking feature status, correct?

Reply:
@patrickdillon I have updated featureSet handling to rely on enabled featureGates based on the featureSet configured. Please let me know if the changes are as expected.